### PR TITLE
Fix scala runner exit codes

### PR DESCRIPTION
**Pull request by rochala**
Adoption of scala2 exit code implementation. Fixes https://github.com/lampepfl/dotty/issues/15022

Now main runner returns proper status code when it fails during runtime.
Also tried to cleanup this class a bit. 

I added tests for all combinations of MainGenericCompiler / Runner that I could think of and found that `-print-tasty` flag was returning wrong exit codes despite errors. I didn't know it that's the best place to put them but i think they count as `scripting` tests.

The last change that I've made is improving error messages for scala expressions (`-e`). Before it created virtual files from `compileFromStrings` even though we explicitly passed one source. It lead to very high horizontal length of stack trace making it less clear as those files had `UUID` in their name. It now creates single file to make it shorter.
```
Exception in thread "main" java.lang.RuntimeException
        at compileFromString$minus86740691$minus6469$minus484b$minusaf4f$minus40ffc988311b$package$.main(compileFromString-86740691-6469-484b-af4f-40ffc988311b.scala:2)
        at main.main(compileFromString-86740691-6469-484b-af4f-40ffc988311b.scala:1)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at dotty.tools.scripting.StringDriver.compileAndRun(StringDriver.scala:36)
        at dotty.tools.MainGenericRunner$.run$1(MainGenericRunner.scala:258)
        at dotty.tools.MainGenericRunner$.main(MainGenericRunner.scala:268)
        at dotty.tools.MainGenericRunner.main(MainGenericRunner.scala)

```

### CODE REVIEW

**Comment by rochala** [comment]
I'll delete this additional file after fixes in code review to not rerun tests again. [touchedFile.out](https://github.com/lampepfl/dotty/pull/15604/files#diff-a2173534ff63b613917c674a28209d762e04f940c6d2e4b6bb8d0e49fbd111da)

File: compiler/test/dotty/tools/scripting/BashExitCodeTests.scala [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    }
+
+  @Test def verifyExitCodeOnScriptCompiler: Unit =
+    assertTestExists("scriptPositive.sc") { file =>
+      testCommandExitCode(Seq(scalacPath, "-script", file.absPath), 0)
+    }
+
+  @Test def verifyExitCodeOnDecompilation: Unit =
+    val testName = "positiveTest"
+    val outputDirectory = testCompilationExitCode(s"$testName.scala", 0)
```
Comment (bishabosha) [inline compiler/test/dotty/tools/scripting/BashExitCodeTests.scala:124]: this doesn't seem to be deleted - same with for `verifyExitCodeOnPrintTasty`

File: compiler/test/dotty/tools/scripting/BashExitCodeTests.scala [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+   *
+   * @param testName name of the test containing the extension
+   * @param test check to be run on found test file
+   */
+  private def assertTestExists(testName: String)(test: File => Unit) =
+    val file = testFiles.find(_.getName == testName)
+    assert(file.isDefined)
+    test(file.get)
+
+class BashExitCodeTests:
```
Comment (griggt) [inline compiler/test/dotty/tools/scripting/BashExitCodeTests.scala:90]:
```suggestion
@Category(Array(classOf[BootstrappedOnlyTests]))
class BashExitCodeTests:
```

These tests depend on the output of `dist/pack` which uses the bootstrapped compiler, but we don't want to force bootstrapping in `test_non_bootstrapped` (see #15513).

You'll need to add an import as well: `import org.junit.experimental.categories.Category`

**Review by griggt — CHANGES_REQUESTED** [review]

File: compiler/test/dotty/tools/scripting/BashExitCodeTests.scala (hunk 2 above)
Comment (rochala) [reply compiler/test/dotty/tools/scripting/BashExitCodeTests.scala:124]: Yeah, thanks for noticing

File: compiler/test/dotty/tools/scripting/BashExitCodeTests.scala (hunk 1 above)
Comment (rochala) [reply compiler/test/dotty/tools/scripting/BashExitCodeTests.scala]: I refactored it to make it simpler to create tests and always guarantee temporary file deletion

File: compiler/test/dotty/tools/scripting/ScriptingTests.scala [hunk 3]
```diff
@@ -146,6 +146,8 @@ class ScriptingTests:
     if touchedFile.exists then
       printf("success: executable jar created file %s\n", touchedFile)
     assert( touchedFile.exists, s"expected to find file ${touchedFile}" )
+    touchedFile.delete
+    assert( !touchedFile.exists, s"unable to delete ${touchedFile}" )
```
Comment (rochala) [inline compiler/test/dotty/tools/scripting/ScriptingTests.scala:150]: I found that touchedFile.out was a leftover from this test. Now it should be deleted properly.

**Review by bishabosha — APPROVED** [review]

**Review by bishabosha — CHANGES_REQUESTED** [review]
minor point but it would be nice if for `ExecuteMode.Expression` the `StringDriver.compileAndRun` was escaped for the `StringDriverException` to just print the message, like what happens for `ScriptingException`.

Currently we get a stack trace for compilation errors in the `-e` mode:
```scala
~/workspace/scripts » ../dotty/dist/target/pack/bin/scala -e 'prinln(2)' || echo "fail"
-- [E006] Not Found Error: expression:2:2 --------------------------------------
2 |  prinln(2)
  |  ^^^^^^
  |  Not found: prinln
  |
  | longer explanation available when compiling with `-explain`
dotty.tools.scripting.StringDriverException: Errors encountered during compilation
	at dotty.tools.scripting.StringDriverException$.apply(StringDriver.scala:52)
	at dotty.tools.scripting.StringDriver.compileAndRun(StringDriver.scala:31)
	at dotty.tools.MainGenericRunner$.run$1(MainGenericRunner.scala:261)
	at dotty.tools.MainGenericRunner$.process(MainGenericRunner.scala:271)
	at dotty.tools.MainGenericRunner$.main(MainGenericRunner.scala:281)
	at dotty.tools.MainGenericRunner.main(MainGenericRunner.scala)
fail
```

**Comment by rochala** [comment]
Yeah, my mistake, and while we're at expression flag. In scala 2 `-e` flag is called as `execute` while here it's called `expression`. Shouldn't scala3 match scala2 naming ?

**Comment by bishabosha** [comment]
> Yeah, my mistake, and while we're at expression flag. In scala 2 `-e` flag is called as `execute` while here it's called `expression`. Shouldn't scala3 match scala2 naming ?

I don't think its a big deal just because its internal details, but could also cause confusion with other modes, such as `ExecuteMode.Run`

File: scoverage.coverage [hunk 4]
```diff
@@ -0,0 +1,88 @@
+# Coverage data, format version: 3.0
```
Comment (bishabosha) [inline scoverage.coverage:1]: possible mistake?

**Review by bishabosha — APPROVED** [review]

**Pull request merged**